rle/rleplus_reader.go

```diff
@@ -4,6 +4,59 @@ import (
 	"golang.org/x/xerrors"
 )
 
+type decodeInfo struct {
+	length byte // length of the run
+	i      byte // i+1 is number of repeats of above run lengths
+	n      byte // number of bits to read
+	varint bool // varint signifies that futher bits need to be processed as a varint
+}
+
+func init() {
+	buildDecodeTable()
+}
+
+// this is a LUT for all possible 6 bit codes and what they decode into
+// possible combinations are:
+// 0bxxxxxx1 - 1 run of 1
+// 0bxxxxx11 - 2 runs of 1
+// up to 0b111111 - 6 runs of 1
+// 0bAAAA10 - 1 run of length 0bAAAA
+// 0bxxxx00 - var int run, the decode value not defined in LUT
+var decodeTable = [1 << 6]decodeInfo{}
+
+func buildDecodeTable() {
+	// runs of 1s 0bxxxxxx1, 0bxxxx11 ...
+	for i := 1; i <= 6; i++ {
+		for j := 0; j < (1<<6)>>i; j++ {
+			idx := bitMasks[i] | byte(j<<i)
+			decodeTable[idx] = decodeInfo{
+				length: 1,
+				i:      byte(i - 1),
+				n:      byte(i),
```

Comment:
I believe this produces a table that works, but it's not the optimal table. For example:

```
i=1
j=1
idx=0b000011
```

We'll end up treating the run as length 1 instead of length 2, right?

Reply:
What if we iterated over all `i in 1<<6` and just decoded these numbers? That seems like a more reliable approach.

Reply:
it will be two repeates (`i` is 0 based) of run length of 1.
The higher number of repeats are set letter so they override the lower number of repeats with more specific decode info.

Iterating over the whole range and decoding makes sense, but I'm not sure what is more complex.

Reply:
Ah, got it. I did not expect this to overwrite.